pdm/models/requirements.py

```diff
@@ -118,9 +118,7 @@ def marker(self, value: Optional[Marker]) -> None:
         except InvalidMarker as e:
             raise RequirementError("Invalid marker: %s" % str(e)) from None
 
-    def identify(self) -> Optional[str]:
-        if not self.key:
-            return None
+    def identify(self) -> str:
```

Comment:
@linw1995

changing this signature results in a failing test

```
Traceback (most recent call last):
  File "/home/runner/work/pdm/pdm/pdm/termui.py", line 192, in logging
    yield logger
  File "/home/runner/work/pdm/pdm/tests/resolver/test_resolve.py", line 46, in resolve_requirements
    mapping, *_ = resolve(resolver, requirements, requires_python)
  File "/home/runner/work/pdm/pdm/pdm/resolver/core.py", line 150, in resolve
    result = resolver.resolve(requirements, max_rounds)
  File "/home/runner/work/pdm/pdm/__pypackages__/3.7/lib/resolvelib/resolvers.py", line 473, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/runner/work/pdm/pdm/__pypackages__/3.7/lib/resolvelib/resolvers.py", line 341, in resolve
    name, crit = self._merge_into_criterion(r, parent=None)
  File "/home/runner/work/pdm/pdm/pdm/resolver/core.py", line 24, in _merge_into_criterion
    identifier, crit = _old_merge_into_criterion(self, requirement, parent)
  File "/home/runner/work/pdm/pdm/__pypackages__/3.7/lib/resolvelib/resolvers.py", line 140, in _merge_into_criterion
    identifier = self._p.identify(requirement_or_candidate=requirement)
  File "/home/runner/work/pdm/pdm/pdm/resolver/providers.py", line 30, in identify
    return requirement_or_candidate.identify()
  File "/home/runner/work/pdm/pdm/pdm/models/requirements.py", line 123, in identify
    return self.key + extras
TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
```

Reply:
https://github.com/danieleades/pdm/blob/refactor/pdm.cli.utils/pdm/models/requirements.py#L309-L312

There has a missing procedure to read the name and version from the source distribution. So the name of the requirement is not set.

Reply:
What's the way forward here?

Reply:
@linw1995 We can expand the logic to extract the name and version from egg info too: `<name>-<version>.tar.gz`

Reply:
`None` is a possible value for `identify()`, when the package is a local directory and its name can't be known until it is built.

Reply:
> `None` is a possible value for `identify()`, when the package is a local directory and its name can't be known until it is built.

if `identify` can return `None`, then i don't think it should be used as a unique key. is there an alternative unique identify that can be used? Would you not get a collision if you have multiple keys which are `None`?

Reply:
> if identify can return None, then i don't think it should be used as a unique key. is there an alternative unique identify that can be used? Would you not get a collision if you have multiple keys which are None?

Good point, we didn't consider that case, a generated random name should be better and we can replace it later.

Reply:
> > if identify can return None, then i don't think it should be used as a unique key. is there an alternative unique identify that can be used? Would you not get a collision if you have multiple keys which are None?
> 
> Good point, we didn't consider that case, a generated random name should be better and we can replace it later.

Maybe use the path of the local directory which is relative to project root as the result of `identify()`.

Reply:
i've updated this to use the path if the key is not set. this is failing one test-

```
_ test_resolve_local_artifacts[/home/runner/work/pdm/pdm/tests/fixtures/artifacts/demo-0.0.1.tar.gz] _
[gw0] linux -- Python 3.7.10 /opt/hostedtoolcache/Python/3.7.10/x64/bin/python

project = <Project '/tmp/pytest-of-runner/pytest-0/popen-gw0/test_resolve_local_artifacts__0'>
repository = <tests.conftest.TestRepository object at 0x7f1d3f6d3d50>
requirement_line = '/home/runner/work/pdm/pdm/tests/fixtures/artifacts/demo-0.0.1.tar.gz'

    @pytest.mark.parametrize(
        "requirement_line",
        [
            f"{(FIXTURES / 'artifacts/demo-0.0.1.tar.gz').as_posix()}",
            f"{(FIXTURES / 'artifacts/demo-0.0.1-py2.py3-none-any.whl').as_posix()}",
        ],
    )
    def test_resolve_local_artifacts(project, repository, requirement_line):
>       result = resolve_requirements(repository, [requirement_line])

tests/resolver/test_resolve.py:109: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/resolver/test_resolve.py:46: in resolve_requirements
    mapping, *_ = resolve(resolver, requirements, requires_python)
pdm/resolver/core.py:153: in resolve
    all_metasets = extract_metadata(result)
pdm/resolver/core.py:107: in extract_metadata
    new_metasets[k] = _build_metaset(crit, all_metasets, keep_unresolved)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

criterion = Criterion((<NamedRequirement chardet; os_name == "nt">, via=<Candidate demo 0.0.1 from None>))
all_metasets = {}, keep_unresolved = set()

    def _build_metaset(
        criterion: Criterion,
        all_metasets: Dict[str, Metaset],
        keep_unresolved: Set[Optional[str]],
    ) -> Metaset:
    
        metaset = None
    
        for r, parent in criterion.information:
            if parent and _identify_parent(parent) in keep_unresolved:
                continue
            this_metaset = Metaset(r.marker)
            # Use 'and' to connect markers inherited from parent.
            if not parent:
                parent_metaset = Metaset()
            else:
>               parent_metaset = all_metasets[_identify_parent(parent)]
E               KeyError: 'demo'

pdm/resolver/core.py:56: KeyError
```